### DISCUSSION TITLE
[RFC] 022 - 环境变量配置默认助手参数

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#913] [RFCs]
## 背景

不同用户对于初始化模型的需求不一样：

- #782 
- #811
- #912
- #1010 
- #1196

因此最好提供一个环境变量来自定义初始化模型的参数

## 实现思路

1. 环境变量填写

参考 PluginSettings 环境变量的实现逻辑，用 `key=value` 的语法允许用户添加需要自定义的配置。级联的字段需要使用 . 来分割

比如同时配置 model 、 params 里的maxToken，以及插件列表

```env
DEFAULT_AGENT_CONFIG='model=gpt-4-1106-preview;params.max_tokens=300;plugins=search-engine,lobe-image-designer'
```

2. 加载配置

目前默认的设定助手是在前端写死的固定值。现在需要从服务端读值后 合并，合并顺序应该为:

1. 前端默认值
2. 服务端设置
3. 用户客户端配置

由于 merge 一定是后来者的配置合并前者，因此现在采用 persist 的方式存在一定的问题，需要改造成异步加载取数模式。

这个改造也可以为实施 #368 做好铺垫

## 补充内容

由于该环境变量依赖了 lobe agent 的数据结构定义。 因此需要补充一篇文档单独介绍 agent 的数据结构定义。结合 https://github.com/lobehub/lobe-chat/discussions/924 考虑

**Comment by arvinxx** [comment]
# SettingStore 存储改造思路

setting 可以直接重构放进 localDB 吗

感觉可以放进去，这样 settings 的迁移就可以使用现成的方案来实现了，并且可以和后续的跨端同步方案保持一致。

那这样就只需要实现有一个 getSettings 和 updateSettings 的service方法

数据迁移可以采用schema升级来自动实现

那这样一来，之前在presist写的merge方法就都不需要了。同时保存时候的检验也能做。

如此一来，在获取settings数据时就可以做一个同步合并，即拿 server default agent + db settings 得到最后的 default agent。

然后用户在存settings时，应该只需要存变更的数据部分，而不是全量 agentSettings。这样就可以细粒度控制需要保存的状态了。

**Comment by arvinxx** [comment]
## Settings 的 JSON Patch 化

重构 settings 存储时遇到一个问题，如果直接使用 简单的 diff 实现，会造成无法正常触发数据保存。一开始的实现如下：
```ts
import { isEqual, isObject, transform } from 'lodash-es';

function difference(object, base) {
    function changes(object, base) {
        return transform(object, (result, value, key) => {
            if (!isEqual(value, base[key])) {
                result[key] = (isObject(value) && isObject(base[key])) ? changes(value, base[key]) : value;
            }
        });
    }
    return changes(object, base);
}

// 使用方法
const settings = {
    theme: 'dark',
    notifications: {
        follow: true,
        alerts: false,
        unfollow: true
    },
    privacy: {
        search: false
    }
};

const defaultSettings = {
    theme: 'light',
    notifications: {
        follow: false,
        alerts: true,
        unfollow: false
    },
    privacy: {
        search: true
    }
};

const modifiedSettings = difference(settings, defaultSettings);
console.log(modifiedSettings);
```


实际使用时，比如默认设置为：
```js
 defaultSettings: {
    fontSize: 14,
    language: 'auto',
    password: '',
    themeMode: 'auto'
}
```

此时，我修改 themeMode 为 `dark`，diff 数据为：
```js
{ themeMode:"dark" }
```
这样没毛病，然后也正常存入数据库。

然后接下来再切回 themeMode 到 `auto`，此时的diff数据就变成了：

```js
{ }
```

那这里就出现了问题，因为 diff 为空， 保存到数据库里的字段不会触发变更（针对 settings 的变动是 deepMerge ）

所以要找另外一种方案来解决只存储变更部分数据的需求。

**Reply by arvinxx** [reply]
### JSON Patch

JSON Patch 是一种标准格式（[RFC 6902](https://datatracker.ietf.org/doc/html/rfc6902)），用来描述 JSON 数据之间的差异并应用这些差异来更新 JSON 文档。它是一种轻量级的操作列表，每个操作代表对 JSON 文档的一种修改，如添加、删除、替换、移动或复制值。

一个 JSON Patch 文档是一个 JSON 数组，每个元素是一个具有特定字段的对象，用来表示单个操作。下面是一些操作的例子：

1. **添加（add）** - 在指定的路径添加一个值：
   ```json
   { "op": "add", "path": "/a/b/c", "value": "foo" }
   ```
   这会在路径 `/a/b/c` 处添加值 `"foo"`。

2. **删除（remove）** - 删除指定路径的值：
   ```json
   { "op": "remove", "path": "/a/b/c" }
   ```
   这会移除路径 `/a/b/c` 处的值。

3. **替换（replace）** - 替换指定路径的值：
   ```json
   { "op": "replace", "path": "/a/b/c", "value": "bar" }
   ```
   这会将路径 `/a/b/c` 处的值替换为 `"bar"`。

4. **移动（move）** - 将一个值从一个路径移动到另一个路径：
   ```json
   { "op": "move", "from": "/a/b/c", "path": "/a/b/d" }
   ```
   这会将值从路径 `/a/b/c` 移动到 `/a/b/d`。

5. **复制（copy）** - 将一个值从一个路径复制到另一个路径：
   ```json
   { "op": "copy", "from": "/a/b/c", "path": "/a/b/e" }
   ```
   这会复制路径 `/a/b/c` 处的值到新的路径 `/a/b/e`。

6. **测试（test）** - 测试指定路径的值是否等于提供的值：
   ```json
   { "op": "test", "path": "/a/b/c", "value": "foo" }
   ```
   这会测试路径 `/a/b/c` 处的值是否为 `"foo"`。

应用 JSON Patch 时，必须按照数组中的顺序执行操作。如果任何操作失败，则整个 Patch 操作应该被视为失败，并且不应该应用任何更改。

JSON Patch 是处理 JSON 数据变更的强有力工具，特别适用于 API 和网络协议，因为它允许以紧凑的格式传输和应用更新，而不需要发送整个 JSON 文档。它也被用于前后端数据同步，以及在应用程序状态管理中记录和重放用户的操作。

**Reply by arvinxx** [reply]
测试了下 jiff 是一个相对来说直接可用的方案： https://github.com/cujojs/jiff 


### 生成 Patch
```ts
import { diff } from 'jiff';

const diffs = diff(defaultSettings, nextSettings);
```

结果示例：
```json
[
  { "op": "test", "path": "/fontSize", "value": 14 },
  { "op": "replace", "path": "/fontSize", "value": 16 },
  { "op": "add", "path": "/neutralColor", "value": "" },
  { "op": "add", "path": "/primaryColor", "value": "" },
  { "op": "test", "path": "/language", "value": "auto" },
  { "op": "replace", "path": "/language", "value": "en-US" },
  { "op": "test", "path": "/themeMode", "value": "auto" },
  { "op": "replace", "path": "/themeMode", "value": "light" },
  { "op": "test", "path": "/password", "value": "" },
  { "op": "remove", "path": "/password" },
  {
    "op": "test",
    "path": "/tts",
    "value": {
      "openAI": {
        "sttModel": "whisper-1",
        "ttsModel": "tts-1"
      },
      "sttAutoStop": true,
      "sttServer": "openai"
    }
  },
  { "op": "remove", "path": "/tts" },
  {
    "op": "test",
    "path": "/tool",
    "value": {
      "dalle": {
        "autoGenerate": false
      }
    }
  },
  { "op": "remove", "path": "/tool" },
  {
    "op": "test",
    "path": "/languageModel",
    "value": {
      "openAI": {
        "OPENAI_API_KEY": "",
        "models": [
          "gpt-3.5-turbo",
          "gpt-3.5-turbo-1106",
          "gpt-3.5-turbo-16k",
          "gpt-4",
          "gpt-4-32k",
          "gpt-4-1106-preview",
          "gpt-4-vision-preview"
        ]
      }
    }
  },
  { "op": "remove", "path": "/languageModel" },
  {
    "op": "test",
    "path": "/defaultAgent",
    "value": {
      "config": {
        "autoCreateTopicThreshold": 2,
        "displayMode": "chat",
        "enableAutoCreateTopic": true,
        "historyCount": 1,
        "model": "abc",
        "params": {
          "frequency_penalty": 0,
          "presence_penalty": 0,
          "temperature": 0.6,
          "top_p": 1
        },
        "plugins": [],
        "systemRole": "",
        "tts": {
          "showAllLocaleVoice": false,
          "sttLocale": "auto",
          "ttsService": "openai",
          "voice": {
            "openai": "alloy"
          }
        }
      },
      "meta": {}
    }
  },
  { "op": "remove", "path": "/defaultAgent" }
]

```

### 使用 Patch

```ts
import { patch } from 'jiff';

const settings = patch(data.settings || [], get().defaultSettings) as GlobalSettings;
```

**Reply by arvinxx** [reply]
事实上不用 deepMerge 也可以…

**Comment by buwanyuanshen** [comment]
这玩意不是放前端更好吗？

**Reply by arvinxx** [reply]
理论上来说是，但现在没有方法同时支持 vercel 和 docker 配置一个变量可以直接在前端生效的。必须要通过后端发配置给前端才行。

**Reply by hqzh** [reply]
太需要前端可配置了

**Reply by arvinxx** [reply]
等这个实现： https://github.com/lobehub/lobe-chat/discussions/1632

**Comment by Wxh16144** [comment]
补充一点：

如果配置了 `DEFAULT_AGENT_CONFIG` 变量，可以通过 http://localhost:3210/api/config 返回的 json 内容来验证是否符合预期。（这是我自己在配置的时候一点经验～

**Reply by Wxh16144** [reply]
<img width="444" alt="image" src="https://github.com/lobehub/lobe-chat/assets/32004925/e8b61d2f-b83a-4b32-995f-02bb7bc640cc">

**Reply by arvinxx** [reply]
https://github.com/lobehub/lobe-chat/pull/2231 在这个 PR 将路由端点改成了  `/trpc/edge/config.getDefaultAgentConfig`

**Comment by RubuJam** [comment]


**Reply by arvinxx** [reply]
这个似乎是 setting 的部分，目前应该配置不了

**Comment by alexallen1** [comment]
使用DEFAULT_AGENT_CONFIG='model=gpt-4-1106-preview;params.max_tokens=300;plugins=search-engine,lobe-image-designer'
yaml: line 12: could not find expected ':'

**Comment by tisfeng** [comment]
你好，默认设置的模型不支持上传图片，请如何通过参数开启 vision ？

**Reply by tisfeng** [reply]
找了一圈文档，不但环境变量里面没找到这个开关，连 UI 上的默认助手都没看到选项，难道没有？

<img width="1427" alt="25451" src="https://github.com/user-attachments/assets/def33e08-6bac-4f16-a308-3efef12112ac">

**Reply by arvinxx** [reply]
默认设置的模型需要你改 model 配置的。 看下ModelList 这一节内容 https://lobehub.com/zh/docs/self-hosting/advanced/model-list

**Reply by tisfeng** [reply]
你好，我参考文档描述，设置 Vercel 环境变量 CUSTOM_MODELS，但还是无法开启 vision，是我哪里弄错了吗？

<img width="1346" alt="84685" src="https://github.com/user-attachments/assets/00b5f03b-71bf-4e54-8245-9100f06ab5c8">

**Reply by arvinxx** [reply]
CUSTOM MODEL 已经没了

**Reply by tisfeng** [reply]
ok，明白了，改成 OPENAI_MODEL_LIST 就好了 https://lobehub.com/zh/docs/self-hosting/environment-variables/model-provider#openai-model-list 。

**Comment by Anivie** [comment]
看到类似的ISSUE，但是维护团队并没有回复，请问维护团队是否有计划制作类似的功能？如果没有的话可以接受这样功能的PR吗？
通过环境变量指定数据文件，为所有用户提供一系列的默认助手（而不是修改那一个默认助手）
期望：通过添加BUILTIN_ASSISTANTS环境变量指向一个目录，读取其中的json作为默认助手为所有用户提供

**Reply by arvinxx** [reply]
目前我们没有精力做这个，但有 PR 的话欢迎，可以先写一个 RFC 出来一起讨论下。

需要同时考虑下在vercel部署和 docker 部署的场景，必须要满足两方部署均可用。

**Comment by skymkmk** [comment]
历史记录数量涉及到 token 消耗量，直接和金钱有关；当前的默认助手参数并无法限制用户调高历史记录数量，能否添加新的环境变量，这个环境变量直接限制用户可调的历史记录的上限，也就是用户可以自由在这个限制下调节历史记录数量，但不可逾越这个值？

**Comment by g-xiaoduan** [comment]
这model目前只能设置成openai的啊……没法设置成其它厂商的
<img width="1083" height="141" alt="image" src="https://github.com/user-attachments/assets/b1175580-5d0d-401d-9a46-fe63d180fd7d" />
<img width="408" height="54" alt="image" src="https://github.com/user-attachments/assets/e375d6b3-39b7-4b8d-ba45-64bc6a1c5911" />

**Reply by SpeedupMaster** [reply]
DEFAULT_AGENT_CONFIG=model=deepseek-chat;provider=deepseek
